tests/validator/validator.cpp

```diff
@@ -2841,3 +2956,465 @@ TEST(Validator, duplicateIdAll)
+    auto validator = libcellml::Validator::create();
+    auto importer = libcellml::Importer::create();
+
+    auto model = parser->parseModel(fileContents("importer/invalid_import_url.cellml"));
+    EXPECT_EQ(size_t(0), parser->issueCount());
+
+    validator->validateModel(model);
+    EXPECT_EQ(size_t(1), validator->issueCount());
+    EXPECT_EQ(errorMessages[0], validator->issue(0)->description());
+    EXPECT_EQ(model->component("c")->importSource(), validator->error(0)->importSource());
+
+    importer->resolveImports(model, resourcePath("importer/"));
+    EXPECT_EQ(size_t(1), importer->errorCount());
+    EXPECT_EQ(errorMessages[1], importer->issue(0)->description());
+
+    validator->validateModel(model);
+    EXPECT_EQ(size_t(1), validator->issueCount());
+    EXPECT_EQ(errorMessages[0], validator->issue(0)->description());
+    EXPECT_EQ(model->component("c")->importSource(), validator->error(0)->importSource());
+}
+
+TEST(Validator, importSecondGenComponentWithInvalidName)
+{
+    const std::string errorMessageValidator =
+        "Imported component 'c' is not valid because:\n"
+        "  -> Component 'c' imported as 'c' from 'invalid_import_url.cellml' has an error:\n"
+        "  -> Import of component 'c' has an invalid URI in the xlink:href attribute.";
+    const std::string errorMessageImporter =
+        "The attempt to resolve imports with the model at '"
+        + resourcePath("importer/")
+        + "i am broken and invalid.cellml' failed: the file could not be opened.";
+
+    auto parser = libcellml::Parser::create();
+    auto validator = libcellml::Validator::create();
+    auto importer = libcellml::Importer::create();
+
+    auto model = parser->parseModel(fileContents("importer/import_invalid_import_url.cellml"));
+    EXPECT_EQ(size_t(0), parser->issueCount());
+
+    validator->validateModel(model);
+    EXPECT_EQ(size_t(0), validator->issueCount());
+
+    importer->resolveImports(model, resourcePath("importer/"));
+    EXPECT_EQ(size_t(1), importer->issueCount());
+    EXPECT_EQ(errorMessageImporter, importer->issue(0)->description());
+    EXPECT_EQ(model->component("c", true), importer->issue(0)->component());
+
+    validator->validateModel(model);
+    EXPECT_EQ(size_t(1), validator->issueCount());
+    EXPECT_EQ(errorMessageValidator, validator->issue(0)->description());
+}
+
+TEST(Validator, highIndexUnitsImport)
+{
+    const std::string errorMessageImporter =
+        "Import of units 'i_am_bad' from '"
+        + resourcePath("importer/") + "units_library.cellml' requires units named 'ps', which relies on child units named 'seconds', which cannot be found.";
+    const std::string errorMessageValidator =
+        "Imported units 'i_am_bad' is not valid because:\n"
+        "  -> Units 'ps' imported as 'i_am_bad' from 'units_library.cellml' has an error:\n"
+        "  -> Units reference 'seconds' in units 'ps' is not a valid reference to a local units or a standard unit type.";
```

Comment:
I would rephrase to:
```
Units 'ps' is imported from units 'i_am_bad' in 'units_library.cellml' which is invalid (units item 'seconds' is not a valid reference to a local units or a standard unit type).
```

(Note that we shouldn't use "units reference" but "units item".)